src/tests/hotreload-integration-test.ts

```diff
@@ -0,0 +1,164 @@
+    return super.loadWasmBinary({implFile: 'build/tests/source/test-module-old.wasm'});
+  }
+
+  clone(): StubWasmLoader {
+    return this;
+  }
+}
+
+function StubPecFactory(loader: Loader) {
+  return (pecId: Id, idGenerator: IdGenerator) => {
+    const channel = new MessageChannel();
+    const pec = new ParticleExecutionContext(channel.port1, pecId, idGenerator, loader);
+    return channel.port2;
+  };
+}
+
+describe('Hot Code Reload for JS Particle', async () => {
+  it('updates model and template', async () =>{
+    const context = await Manifest.parse(`
+      particle A in 'A.js'
+        consume root
+
+      recipe
+        slot 'rootslotid-root' as slot0
+        A
+          consume root as slot0`);
+
+    const loader = new StubLoader({
+      'A.js': `defineParticle(({DomParticle}) => {
+        return class extends DomParticle {
+          get template() { return 'Hello <span>{{name}}</span>, old age: <span>{{age}}</span>'; }
+
+          render() {
+            return {name: 'Jack', age: '10'};
+          }
+        };
+      });`
+    });
+
+    const newLoader = new StubLoader({
+      'A.js': `defineParticle(({DomParticle}) => {
+        return class extends DomParticle {
+          get template() { return 'Hello <span>{{name}}</span>, new age: <span>{{age}}</span>'; }
+
+          render() {
+            return {name: 'Jack', age: '15'};
+          }
+        };
+      });`
+    });
+
+    const id = ArcId.newForTest('demo');
+    const pecFactories = [StubPecFactory(loader).bind(null)];
+    const storageKey = 'volatile://' + id.toString();
+    const slotComposer = new StubSlotComposer();
+    
+    const arc = new Arc({id, pecFactories, storageKey, slotComposer, loader, context});
+
+    const [recipe] = arc.context.recipes;
+    assert.isTrue(recipe.normalize() && recipe.isResolved());
+    await arc.instantiate(recipe);
+
+    const slotConsumer = slotComposer.consumers[0] as HeadlessSlotDomConsumer;
+
+    const oldModel = slotConsumer.getRendering().model;
+    const oldTemplate = slotConsumer._content.template;
+    assert.deepEqual(oldModel,  {name: 'Jack', age: '10'});
+    assert.deepEqual(oldTemplate, `Hello <span>{{name}}</span>, old age: <span>{{age}}</span>`);
+
+    Object.assign(arc._loader, newLoader);
+    arc.pec.reload(arc.pec.particles);
+    await arc.idle;
+
+    const newModel = slotConsumer.getRendering().model;
+    const newTemplate = slotConsumer._content.template;
+
+    assert.deepEqual(newModel,  {name: 'Jack', age: '15'});
+    assert.deepEqual(newTemplate, `Hello <span>{{name}}</span>, new age: <span>{{age}}</span>`);
+  });
+});
+
+describe('Hot Code Reload for WASM Particle', async () => {
+  before(function() {
+    if (!global['testFlags'].enableWasm) {
+      this.skip();
+    }
+  });
+
+  it('updates model and template', async () => {
+    const loader = new StubWasmLoader();
+    const manifestString = `
+      particle HotReloadTest in 'build/tests/source/test-module-old.wasm'
```

Comment:
I would suggest using just `test-module.wasm`. Using `test-module-old.wasm` is likely to be confusing for future readers, since it's not all that clear from the code here how `test-module-new.wasm` gets mapped in on a reload. Using a name that doesn't match either built module is a clue that something else is dealing with that.

A comment would make it even more clear: `// StubWasmLoader returns test-module-old.wasm or test-module-new.wasm based on the reloaded flag`

Reply:
Updated!